src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -1391,20 +1392,229 @@ static bool nic_already_used(const struct fi_info *prov, struct fi_info **others
     return false;
 }
 
+/* Return the parent object (typically socket) of the NIC */
+static MPIR_hwtopo_gid_t get_nic_parent(struct fi_info *info)
+{
+    if (info->nic->bus_attr->bus_type == FI_BUS_PCI) {
+        struct fi_pci_attr pci = info->nic->bus_attr->attr.pci;
+        return MPIR_hwtopo_get_dev_parent_by_pci(pci.domain_id, pci.bus_id, pci.device_id,
+                                                 pci.function_id);
+    }
+    return MPIR_hwtopo_get_obj_by_name(info->domain_attr->name);
+}
```

Comment:
There are so much nic related code, can we create a separate source file, e.g. `ofi_nic.c`?

Reply:
I think the only big chunks of new code here that could be split out are all static functions that we probably wouldn't want to split out anyway, right? All of this code is used by the actual initialization functions.

Reply:
> I think the only big chunks of new code here that could be split out are all static functions that we probably wouldn't want to split out anyway, right?

Why? There is no performance concerns (if we worry about potential compiler optimizations). If you worry about namespaces, use a private header, e.g. `ofi_nic.h`.